### ISSUE TITLE
At least on Linux, starting "gsc" and then doing "ctrl+z" frequently leads to infinite loop in gsc

### ISSUE THREAD

**Issue by m-i-k-a-e-l** [feeley/gambit#104]
On Debian stable, I type "gsc", enter. Gambit starts. In a separate terminal I start "strace -p" for the newly started Gambit process. Then back in the Gambit terminal, I press ctrl+z.

The Gambit process is now eating some CPU because it goes into an infinite recursion.

strace outputs the log pasted below - web posts show that the ERESTARTNOHAND  (and ERESTARTSYS?) shown by strace, may not actually reach Gambit, but, some kind of error code clearly does reach Gambit.

What we see here is that Gambit's handling of the situation is not harmonious with the OS' expectation - it looks like the OS tries to do some "resume" related work regarding the read call by sending Gambit a signal, but Gambit doesn't respond properly.

This issue should be checked also on an alternative OS.

Strace output:

Process 16690 attached - interrupt to quit
select(5, [4], [], [], NULL)            = ? ERESTARTNOHAND (To be restarted)
--- SIGTSTP (Stopped) @ 0 (0) ---
--- SIGTSTP (Stopped) @ 0 (0) ---
select(5, [4], [], [], NULL)            = 1 (in [4])
setitimer(ITIMER_VIRTUAL, {it_interval={0, 10000}, it_value={0, 10000}}, NULL) = 0
getitimer(ITIMER_VIRTUAL, {it_interval={0, 12000}, it_value={0, 16001}}) = 0
rt_sigaction(SIGVTALRM, {0x543bfc, [], SA_RESTORER|SA_INTERRUPT, 0x7f16314744f0}, NULL, 8) = 0
write(3, "read len=128\n", 13)          = 13
read(4, 0x2f24be4, 128)                 = ? ERESTARTSYS (To be restarted)
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
read(4, 0x2f24be4, 128)                 = ? ERESTARTSYS (To be restarted)
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
read(4, 0x2f24be4, 128)                 = ? ERESTARTSYS (To be restarted)
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
read(4, 0x2f24be4, 128)                 = ? ERESTARTSYS (To be restarted)
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
[continues similar pattern infinitely.]

**Comment by m-i-k-a-e-l** [comment]
(Only for reference, after having done "bg" there also seems to be a problem where Linux bombs gsc with the following signals and Gambit does this ioctl

--- SIGTTOU (Stopped (tty output)) @ 0 (0) ---
--- SIGTTOU (Stopped (tty output)) @ 0 (0) ---
ioctl(4, SNDCTL_TMR_START or TCSETS, {B38400 opost isig icanon echo ...}) = ? ERESTARTSYS (To be restarted)

)

**Comment by m-i-k-a-e-l** [comment]
Similarly to the first post above, if I start a "gsc" and send a SIGINT signal from another terminal ("kill -SIGINT pid"), I get an almost equivalent failure:

(Sending SIGSTOP does *not* trig the bug though.)


Process 16706 attached - interrupt to quit
select(5, [4], [], [], NULL)            = ? ERESTARTNOHAND (To be restarted)
--- SIGSTOP (Stopped (signal)) @ 0 (0) ---
--- SIGSTOP (Stopped (signal)) @ 0 (0) ---
select(5, [4], [], [], NULL)            = ? ERESTARTNOHAND (To be restarted)
--- SIGINT (Interrupt) @ 0 (0) ---
rt_sigreturn(0)                         = -1 EINTR (Interrupted system call)
setitimer(ITIMER_VIRTUAL, {it_interval={0, 10000}, it_value={0, 10000}}, NULL) = 0
getitimer(ITIMER_VIRTUAL, {it_interval={0, 12000}, it_value={0, 16001}}) = 0
rt_sigaction(SIGVTALRM, {0x543bfc, [], SA_RESTORER|SA_INTERRUPT, 0x7f25188eb4f0}, NULL, 8) = 0
write(3, "*** OS ERROR AT \"os_io.c\"@1069.1"..., 70) = 70
write(3, "read len=128\n", 13)          = 13
read(4, 0x2e21be4, 128)                 = ? ERESTARTSYS (To be restarted)
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
read(4, 0x2e21be4, 128)                 = ? ERESTARTSYS (To be restarted)
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
read(4, 0x2e21be4, 128)                 = ? ERESTARTSYS (To be restarted)
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---
--- SIGTTIN (Stopped (tty input)) @ 0 (0) ---

**Comment by feeley** [comment]
To test this out I installed Debian and I can reproduce the problem by starting "strace -p <pid-of-gsc>", then CTRL-Z gsc, then *return* in the shell terminal window where gsc is running.  Then strace generates repeated ERESTARTSYS and SIGTTIN.  This causes gsc (and strace) to eat up CPU time.

However, when strace is terminated (CTRL-C), then gsc stops eating up CPU time.  So it seems that the problem is related to strace that is "seeing" internal ERESTARTSYS events and causing the read system call to be restarted internally, ad infinitum (the read system call doesn't return).

I googled and found a few other softwares behaving this way when being observed with strace. So I'm inclined to blame the issue on a faulty implementation of the OS and/or strace.
